app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/service.tsx

```diff
@@ -10,20 +10,107 @@ import * as loggerProvider from "../providers/logger";
+};
+
 /** Relative HTTP path to the "get user" endpoint of the Cloud backend API. */
 const GET_USER_PATH = "users/me";
+/** Relative HTTP path to the "list projects" endpoint of the Cloud backend API. */
+const LIST_PROJECTS_PATH = "projects";
+/** Relative HTTP path to the "create project" endpoint of the Cloud backend API. */
+const CREATE_PROJECT_PATH = "projects";
+/** Relative HTTP path to the "close project" endpoint of the Cloud backend API. */
+const closeProjectPath = (projectId: ProjectId) =>
+  `projects/${projectId}/close`;
+/** Relative HTTP path to the "get project" endpoint of the Cloud backend API. */
+const getProjectPath = (projectId: ProjectId) => `projects/${projectId}`;
+/** Relative HTTP path to the "open project" endpoint of the Cloud backend API. */
+const openProjectPath = (projectId: ProjectId) => `projects/${projectId}/open`;
 
 // =============
 // === Types ===
 // =============
 
+/** Unique identifier for a user's project. */
+export type ProjectId = string;
+
 /** A user/organization in the application. These are the primary owners of a project. */
 export interface Organization {
   id: string;
   userEmail: string;
   name: string;
 }
 
+/** Type of application that a {@link Version} applies to.
+ *
+ * We keep track of both backend and IDE versions, so that we can update the two independently.
+ * However the format of the version numbers is the same for both, so we can use the same type for
+ * both. We just need this enum to disambiguate. */
+export enum VersionType {
+  backend = "Backend",
+  ide = "Ide",
+}
+
+/** A version describing a release of the backend or IDE. */
+export interface Version {
+  versionType: VersionType;
+  ami: string | undefined;
+  created: string;
+  /** This field name does not follow the naming convention. This field name is snake case on the
+   * backend. The case must match here for JSON deserialization to work. */
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  version_number: string;
+}
+
+/** Possible states that a project can be in. */
+export enum ProjectState {
+  created = "Created",
+  new = "New",
+  openInProgress = "OpenInProgress",
+  opened = "Opened",
+  closed = "Closed",
+}
+
+/** Wrapper around a project state value. */
+export interface ProjectStateType {
+  type: ProjectState;
+}
+
+/** A user/organization's project containing and/or currently executing code. */
+export interface Project {
+  organizationId: string;
+  projectId: ProjectId;
+  name: string;
+  state: ProjectStateType;
+  packageName: string;
+  address: string | null;
+  ami: string | null;
+  ideVersion: Version | null;
+  engineVersion: Version | null;
+}
+
+// =================
+// === Endpoints ===
+// =================
+
+/** HTTP response body for the "list projects" endpoint. */
+interface ListProjectsResponseBody {
+  projects: Project[];
+}
+
+/** HTTP request body for the "create project" endpoint. */
+export type CreateProjectRequestBody = {
```

Comment:
should probably be consistent - either every object type should be an `interface`, or every object type should be a `type`

Reply:
ah of course... maybe you could change `Json` to `object` - or replace `Json` with `object` entirely

it doesn't *quite* mean the same thing, but it should be a good fit (`object` is more or less `typeof x === "object"`)




ps: `type T = { object type here }` is designed to implicitly extend `Record<string, unknown>` - probably so that `{ a: boolean; }` is compatible with `foo<T extends Record<string, boolean>>()`
(surprisingly, you can access arbitrary properties on `Record<string, unknown>` but not on `{}`: [playground example](https://www.typescriptlang.org/play?#code/DYUwLgBAhgXBDeBfCBeBiDcAoKA6K2okARnAEogDGA9gE4AmAPAM5i0CWAdgOYA0EAV04BrTtQDunAHypoGCAHoFEMRDABPAA4gIIWrTqz2zCDQC2mqGHbFQWYvmxcwegGZRKOgKIWN6LESmcD6afmhI2MSylBhAA))

Reply:
Dashboard backend methods extracted to their own PR, so this is no longer relevant.